torchrl/objectives/dqn.py

```diff
@@ -95,8 +96,15 @@ def forward(self, input_tensordict: TensorDictBase) -> TensorDict:
             buffers=self.value_network_buffers,
         )
 
+        action = tensordict.get("action")
         pred_val = td_copy.get("action_value")
-        pred_val_index = (pred_val * action).sum(-1)
+
+        if _CATEGORICAL_ACTION_ENCODING:
```

Comment:
This is a bit tricky

What if someone explicitly builds a policy that uses categorical actions, but does not set the env variable?

My view of the env variable is more along the line of "a way of easily change a whole training script from categorical to one hot and vice-versa".

But the reversed dependency should not hold, and the environment variable should only be checked when building the network.

One option could be to infer the type of action by looking at the `value_network`: if it has registered the action spec, then we can use it to infer the type of action we will see. Otherwise, we must ask to the user to tell us (via an arg in the constructor) what kind of action we should be expecting.

Open to other suggestions obviously :)

Reply:
> What if someone explicitly builds a policy that uses categorical actions, but does not set the env variable?

This is exactly what I meant when mentioned that the current solution couples things a bit too much.

Since actions are passed as pure tensors (even though, wrapped in TensorDict) it is either assumed or configured separately for each class what to do with `action` tensors.

Now I can see two options what we can do here:

1. Whenever we deal with actions, we either pass `action_space` in the constructor or derive it from some other entity (e.g. `value_network` as you proposed). However here we rely on careful configuration of all modules together (in theory, somebody could explicitly configure `value_network` incompatible with an environment). Another drawback is that we "spreading out" the logic for each case across the code base which can make code a harder to understand (I think in case of one-hot/categorical it is fine, but for future approaches it can blow up).
2. We can introduce something like `ActionSpec` which would be an extended version of existing specs, though purely for action processing (deriving actions from `value_network`, etc.). We instantiate the `ActionSpec` once and pass its instance everywhere we need encapsulating work with actions. Although, it is hard for me to estimate what interface `ActionSpec` needs in order to cover all possible cases (I mean globally for all RL cases). And this also would require larger scale refactoring.

tldr; I think 1st option seems better choice for this situation and somewhat follows current code base practices (e.g. `QValueHook`)